First Version

lucasceballos opened the pull request.

Commit:
Feat: first push in this repo
Add code, test, linting, etc.

juanem1 committed:
docs: Fix readme styles

Fix docs style

lucasceballos commented on README.md:
```diff
@@ -1 +1,50 @@
+});
+
+// Aggregate object
+aggregator.aggregate({some: 'object});
```
Estaria bueno sumar otro metodo que sea "aggregateMany" que en lugar de recibir UN objeto reciba un array de objetos y simplemente loopee y llame al metodo "aggregate" asi no lo tenemos que hacer desde afuera.

juanem1 replied:
Agregado

lucasceballos commented on spec/mocks/stream_10.js:
```diff
@@ -0,0 +1,15 @@
+ * Mock generated with https://www.mockaroo.com/
+ */
+module.exports = [
+    { "id": 1, "campaingId": 16, "adId": 127, "city": "Kovači", "impressions": 1, "clicks": 2, "spent": 0.93, "ip": "49.26.161.43" },
```
Como vimos en el call estaria bueno agregar objetos con valor extremo, muy muy muy grandes, muy muy muy chicos, ceros, y negativos.

Sabemos que son escenarios que PROBABLEMENTE no se van a dar, pero no podemos estaro 100% seguros de manera que seria ideal poder estar seguros de que esto funciona con cualquier numero que metamos.

juanem1 replied:
Agregado

lucasceballos reviewed (commented):
Como dijimos, fuera de lo que es review linea a linea de codigo lo unico que tenemos duda es...

¿podemos ejecutar multiples instancias de este aggregator con diferentes configs sin que interfieran entre si? ¿o funcionan tipo singleton?

Lo que necesitamos es poder tener varios en un mismo proceso y que sean independientes entre si.

Commit:
feat: multiple instances.
Add posibility to run multiple instances in the same execution.

Commit:
feat: Fix PR comments
Agrego mas tests, resuelvo tema con decimales, actualizo documentacion.

juanem1 commented:
@lucasceballos agregue los fix y features que hablamos. Calculo que esta para review

juanem1 commented:
A parte corregi el codigo para que se puedan hacer multiples instancias de la libreria en la misma ejecucion.

lucasceballos commented on spec/mocks/edge_cases.js:
```diff
@@ -0,0 +1,15 @@
+ * Mock generated with https://www.mockaroo.com/
+ */
+module.exports = [
+    { "id": 1, "campaingId": 16, "adId": 1, "city": "Kovači", "impressions": 0, "clicks": 0, "spent": 0, "ip": "49.26.161.43" },
```
¿Extremos para arriba no testeamos? es decir, numeros MUY GRANDES.

Por otra parte, 0.000001 son solo cinco ceros despues de la coma, eso es tranquilamente lo que podria costar UNA impresion. ¿te parece si lo extremamos aun mas? tipo, 10 o 15 ceros, quiero intentar romperlo, ver que pasa con los redondeos, etc. ¿que opinas?

lucasceballos reviewed (commented):
Veo muchos cambios con respecto a la anterior version, ¿pasar de singleton a multiple instancia fue muy complejo?

Algunos comentarios e ideas que tengo mas alla de lo que deje especificamente en el codigo

1. Veo mucho loop por todos lados, estamos usando siempre for, ¿que tan asincrono es eso? ¿es mejor si usamos async?

2 - En funcion del item anterior, me preocupa un poco la performance de esto, ¿que tan rapido funciona? ¿que tan performante es? creo que estaria bueno agregar un mini script de bench que pruebe alguna agregacion semi heavy y real sobre un dataset grande, te dejo aca adjunto un zip con varios miles de clicks viejos, ¿te parece bien agregar algo asi?


[xaa.zip](https://github.com/HwdGroup/hwd-aggregator/files/645636/xaa.zip)


3 - Asigno a @pablitovicente que es el otro experto en performance nodejs para que nos de su opinion y haga otro review (ademas yo no puedo hacer approves o requests changes porque abri el PR yo).

juanem1 commented:
@lucasceballos No, cambiar de singleton a multime instancia solo me llevo cambiar una linea (module.exports = function ()) al comienzo del archivo.
En cuanto a los for, es el bucle mas performante de javascript asi que no creo que tengamos problemas.

juanem1 commented:
@lucasceballos Ahora hago un test con el archivo que me pasaste, veo los benchmark y te paso los resultados. No creo que sea muy útil ya que no tenemos una base line anterior contra que comparar, pero para sacarnos la duda.

juanem1 commented:
@pablitovicente Avisame cuando lo puedas ver, y si queres hacemos una mini call y te explico como viene la mano.

lucasceballos commented:
Si claro, no tenemos una metrica de referencia, pero es bueno generar ese test para poder ir viendo los diferentes cambios que metemos si mejoran o no.

Por ejemplo podriamos usarlo para ver si async mejora el rendimiento o no.

Commit:
feat: agrego benchmark
Agrego archivo de benchmark para testear resultados en base a un gran archivo de datos.

juanem1 commented:
@lucasceballos  @pablitovicente  Agregue un archivo de benchmark, acá paso los resultados. 
para un archivo de 95MB 100k json agrupando por 8 dimensiones:
node 0.12 tarda 17 segundos
node 6.9   tarda 13 segundos

Commit:
perf: Uso nativo de MD5
Por defecto nodejs tiene una libreria para generar hash md5, lo que nos da mucha mas performance que usando una libreria externa.

juanem1 commented:
Me di cuenta que nodejs ya trae integrada una libreria para generar hash md5. Usando la libreria MD5 nativa el benchmark queda asi. File 95MB 100k json
node 0.12 tarda 12 segundos (29 % de improvement)
node 9.9   tarda   9 segundos (30 % de improvement)

Documentacion del modulo crypto para node 0.12: [https://nodejs.org/docs/latest-v0.12.x/api/crypto.html](https://nodejs.org/docs/latest-v0.12.x/api/crypto.html)

leamarty commented on README.md:
```diff
@@ -1 +1,83 @@
+});
+
+// Aggregate single object
+aggregator.aggregate({some: 'object});
```
Falta una comilla después de ````object````

leamarty commented on README.md:
```diff
@@ -1 +1,83 @@
+// Other option is batch aggregation
+var array = [
+  {some: 'object}, 
+  {other: 'object}
```
Lo mismo en estas dos lineas

Commit:
fix: fix PR comments
Agregue las comillas que faltaban en el readme.

leamarty commented on src/index.js:
```diff
@@ -0,0 +1,212 @@
+   */
+  function makeHash(record) {
+    var objToHash = {};
+    for (var item in record) {
```
Hay una función de underscore que se llama ````_.pick( )```` que se puede usar para esto...

juanem1 replied:
Si, lo se, pero meter una libreria que baja la performance general del script por un for de 3 lineas no vale la pena.
En todo caso nos conviene usar lodash que es mucho mas performante, aunque en este caso no vale la pena. (para usar lodash tenemos que usar node 4+)

leamarty replied:
Es verdad. Que quede así.

leamarty commented on README.md:
````diff
@@ -1 +1,83 @@
+After aggregate and grouping by campaignId you will get:
+```
+{
+  254a86bfca1f75c6ac8541f866707e90: {impressions:10, spent:0.30}
````
Por lo que veo en el código también te devuelve el "campaignId", ¿o no?

juanem1 replied:
Si devuelve todos los valores del grupBy. No lo agregue en la documentación para que sea mas legible, pero si te parece imprescindible los agrego sin problema.

leamarty replied:
Como quieras, yo lo agregaría.

juanem1 replied:
Al final lo agregue y cambie un poquito el ejemplo para que se entienda mejor :smile:

Commit:
fix: Agrego detalle al readme
Agrego detalle al readme de los valores que entrega el script
